src/bsp/src/esp/src/SpiStmMock.cpp

```diff
@@ -28,43 +23,73 @@ SpiStm::SpiStm(ILogger& logger) :
     m_outboundMessage = {0};
     m_inboundHeader = {0};
     m_outboundHeader = {0};
-    m_isBusy = false;
+    m_isConnected = false;
+    m_crcOK = false;
+
+    CircularBuff_init(&m_circularBuf, m_data.data(), m_data.size());
 
     setCallback(SpiStm::transactionCallback, this);
 
     m_driverTask.start();
 }
 
-bool SpiStm::send(const uint8_t* buffer, uint16_t length) {
+bool SpiStm::receive(uint8_t* data, uint16_t length) {
+    if (data == nullptr || length > STM_SPI_MAX_MESSAGE_LENGTH) {
+        m_logger.log(LogLevel::Warn, "Invalid parameters for SpiStm::Receive");
+        return false;
+    }
+    m_receivingTaskHandle = xTaskGetCurrentTaskHandle();
+    while (CircularBuff_getLength(&m_circularBuf) < length) {
+        ulTaskNotifyTake(pdTRUE, 500);
+        // TODO: check for disconnection
+    }
+    m_receivingTaskHandle = nullptr;
+
+    return CircularBuff_get(&m_circularBuf, data, length) == length;
+}
 
-    if (isBusy()) {
-        return false; // Not available
-    } else if (length >= STM_SPI_MAX_MESSAGE_LENGTH) { // Message too long
+bool SpiStm::send(const uint8_t* buffer, uint16_t length) {
+    if (length >= STM_SPI_MAX_MESSAGE_LENGTH) { // Message too long
         m_logger.log(LogLevel::Error,
                      "StmEsp: Message length of %d is larger than maximum allowed of %d", length,
                      STM_SPI_MAX_MESSAGE_LENGTH);
         return false;
-    } else {
-        m_logger.log(LogLevel::Debug, "Sending message of length %d", length);
-        // Memcpy necessary to have buffer word-alligned for transfer
-        std::memcpy(m_outboundMessage.m_data.data(), buffer, length);
-        // Padding with 0 up to a word-alligned boundary
-        for (uint8_t i = 0; i < (length % 4); i++) {
-            m_outboundMessage.m_data[length] = 0;
-            length++;
-        }
-        // Appending CRC32
-        *(uint32_t*)(m_outboundMessage.m_data.data() + length) =
-            calculateCRC32_software(buffer, length);
-        length += CRC32_SIZE;
-        m_outboundMessage.m_sizeBytes = length;
-        m_isBusy = true;
-        notifyMaster();
-        return true;
     }
+
+    m_logger.log(LogLevel::Debug, "Sending message of length %d", length);
+    // Memcpy necessary to have buffer word-alligned for transfer
+    std::memcpy(m_outboundMessage.m_data.data(), buffer, length);
+    // Set payload size
+    m_outboundMessage.m_payloadSizeBytes = length;
+    // Padding with 0 up to a word-alligned boundary
+    while (length % 4 != 0) {
+        m_outboundMessage.m_data[(uint16_t)(length)] = 0;
+        length++;
+    }
```

Comment:
Is it really necessary to set the data to 0 or it could be possible to juste just do length += length%4, anyway the alignement bytes are not considered on the other side right?

Reply:
We need to set it to zero to make sure the crcs match on both side